File: implementations/ios-sdk/uitests/Tests/LiveUpdatesTests.swift
import XCTest

/// Pattern that the `*-entry-id` Text nodes render via LiveUpdatesEntryDisplay.
/// The id segment after "Entry: " is a Contentful sys.id (alphanumeric, no spaces).
/// Asserting against this proves the SDK actually resolved an entry rather than
/// rendering an empty/default state.
private let ENTRY_ID_TEXT_PATTERN = #/^Entry: [a-zA-Z0-9]+$/#

final class LiveUpdatesTests: XCTestCase {
    let app = XCUIApplication()

    override func setUp() {
        continueAfterFailure = false
        app.launch()
        clearProfileState(app: app)

        waitForElement(app.buttons["live-updates-test-button"])
        app.buttons["live-updates-test-button"].tap()

        waitForElement(app.otherElements["default-optimization"])
    }

    override func tearDown() {
        let closeButton = app.buttons["close-live-updates-test-button"]
        if closeButton.exists { closeButton.tap() }

        let mainButton = app.buttons["live-updates-test-button"]
        _ = mainButton.waitForExistence(timeout: ELEMENT_VISIBILITY_TIMEOUT)
    }

    // MARK: - Default behavior (locked on first value)

    func testDefaultDoesNotUpdateOnIdentifyGlobalLiveUpdatesFalse() {
        // Capture the SDK-resolved entry id for both the locked default section
        // and the always-live reference section.
        waitForElement(app.staticTexts["default-entry-id"])
        waitForElement(app.staticTexts["live-entry-id"])

        let initialDefaultEntryIdText = getElementTextById("default-entry-id", app: app)
        let initialLiveEntryIdText = getElementTextById("live-entry-id", app: app)

        app.buttons["live-updates-identify-button"].tap()
        waitForTextEquals("identified-status", expected: "Yes", app: app)

        // The live-prefixed section has liveUpdates=true, so it MUST re-resolve
        // to a different variant after identify. Without this proof, the
        // "locked stays locked" assertion below would be meaningless.
        waitForTextChange("live-entry-id", baseline: initialLiveEntryIdText, app: app)

        // Default section inherits the global setting (off), so the lock must hold.
        waitForTextEquals("default-entry-id", expected: initialDefaultEntryIdText, app: app)
    }

    // MARK: - Global liveUpdates enabled

    func testGlobalLiveUpdatesEnablesDefaultComponents() {
        app.buttons["toggle-global-live-updates-button"].tap()
        waitForTextEquals("global-live-updates-status", expected: "ON", app: app)

        waitForElement(app.staticTexts["default-entry-id"])
        let initialDefaultEntryIdText = getElementTextById("default-entry-id", app: app)

        app.buttons["live-updates-identify-button"].tap()
        waitForTextEquals("identified-status", expected: "Yes", app: app)

        waitForTextChange("default-entry-id", baseline: initialDefaultEntryIdText, app: app)
    }

    func testLockedComponentsIgnoreGlobalLiveUpdates() {
        app.buttons["toggle-global-live-updates-button"].tap()
        waitForTextEquals("global-live-updates-status", expected: "ON", app: app)

        waitForElement(app.staticTexts["locked-entry-id"])
        waitForElement(app.staticTexts["default-entry-id"])

        let initialLockedEntryIdText = getElementTextById("locked-entry-id", app: app)
        let initialDefaultEntryIdText = getElementTextById("default-entry-id", app: app)

        app.buttons["live-updates-identify-button"].tap()
        waitForTextEquals("identified-status", expected: "Yes", app: app)

        // With global=ON the default section (no per-component prop) MUST
        // re-resolve. This is the live-reference that proves the SDK is
        // actually swapping variants.
        waitForTextChange("default-entry-id", baseline: initialDefaultEntryIdText, app: app)

        // Locked section has liveUpdates=false, so it must stay at its captured id.
        waitForTextEquals("locked-entry-id", expected: initialLockedEntryIdText, app: app)
    }

    // MARK: - Per-component liveUpdates=true

    func testLiveComponentUpdatesRegardlessOfGlobal() {
        waitForTextEquals("global-live-updates-status", expected: "OFF", app: app)
        waitForElement(app.staticTexts["live-entry-id"])

        let initialLiveEntryIdText = getElementTextById("live-entry-id", app: app)

        app.buttons["live-updates-identify-button"].tap()
        waitForTextEquals("identified-status", expected: "Yes", app: app)

        waitForTextChange("live-entry-id", baseline: initialLiveEntryIdText, app: app)
    }

    // MARK: - Per-component liveUpdates=false

    func testLockedComponentDoesNotUpdateEvenWhenGlobalOn() {
        app.buttons["toggle-global-live-updates-button"].tap()
        waitForTextEquals("global-live-updates-status", expected: "ON", app: app)

        waitForElement(app.staticTexts["locked-entry-id"])
        waitForElement(app.staticTexts["live-entry-id"])

        let initialLockedEntryIdText = getElementTextById("locked-entry-id", app: app)
        let initialLiveEntryIdText = getElementTextById("live-entry-id", app: app)

        app.buttons["live-updates-identify-button"].tap()
        waitForTextEquals("identified-status", expected: "Yes", app: app)

        // Live section (per-component liveUpdates=true) MUST change — the SDK is
        // re-resolving on identify. The per-component prop is the path under
        // test: it must override the global=true setting and keep the locked
        // section stable.
        waitForTextChange("live-entry-id", baseline: initialLiveEntryIdText, app: app)
        waitForTextEquals("locked-entry-id", expected: initialLockedEntryIdText, app: app)
    }

    // MARK: - Preview panel simulation

    func testPreviewPanelEnablesLiveUpdatesForAll() {
        waitForTextEquals("preview-panel-status", expected: "Closed", app: app)
        app.buttons["simulate-preview-panel-button"].tap()
        waitForTextEquals("preview-panel-status", expected: "Open", app: app)

        waitForElement(app.staticTexts["default-entry-id"])
        waitForElement(app.staticTexts["live-entry-id"])
        waitForElement(app.staticTexts["locked-entry-id"])

        let initialDefaultEntryIdText = getElementTextById("default-entry-id", app: app)
        let initialLiveEntryIdText = getElementTextById("live-entry-id", app: app)
        let initialLockedEntryIdText = getElementTextById("locked-entry-id", app: app)

        app.buttons["live-updates-identify-button"].tap()
        waitForTextEquals("identified-status", expected: "Yes", app: app)

        // While the preview panel is open, the SDK forces shouldLiveUpdate=true
        // for ALL sections, including the per-component liveUpdates=false one.
        // All three resolved variants must change.
        waitForTextChange("default-entry-id", baseline: initialDefaultEntryIdText, app: app)
        waitForTextChange("live-entry-id", baseline: initialLiveEntryIdText, app: app)
        waitForTextChange("locked-entry-id", baseline: initialLockedEntryIdText, app: app)
    }

    // MARK: - Screen controls

    func testToggleGlobalLiveUpdates() {
        waitForTextEquals("global-live-updates-status", expected: "OFF", app: app)
        app.buttons["toggle-global-live-updates-button"].tap()
        waitForTextEquals("global-live-updates-status", expected: "ON", app: app)
        app.buttons["toggle-global-live-updates-button"].tap()
        waitForTextEquals("global-live-updates-status", expected: "OFF", app: app)
    }

    func testTogglePreviewPanel() {
        waitForTextEquals("preview-panel-status", expected: "Closed", app: app)
        app.buttons["simulate-preview-panel-button"].tap()
        waitForTextEquals("preview-panel-status", expected: "Open", app: app)
        app.buttons["simulate-preview-panel-button"].tap()
        waitForTextEquals("preview-panel-status", expected: "Closed", app: app)
    }

    func testIdentifyAndReset() {
        waitForTextEquals("identified-status", expected: "No", app: app)
        app.buttons["live-updates-identify-button"].tap()
        waitForTextEquals("identified-status", expected: "Yes", app: app)
        app.buttons["live-updates-reset-button"].tap()
        waitForTextEquals("identified-status", expected: "No", app: app)
    }

    // MARK: - Three Optimization sections display

    func testDisplaysAllThreeOptimizationEntrySections() {
        waitForElement(app.otherElements["default-optimization"])
        waitForElement(app.otherElements["live-optimization"])
        waitForElement(app.otherElements["locked-optimization"])

        // Section wrappers render unconditionally — proving they're mounted is
        // just a smoke check. The SDK responsibility is to feed each section a
        // resolved entry whose sys.id surfaces in the entry-id Text.
        waitForElement(app.staticTexts["default-entry-id"])
        waitForElement(app.staticTexts["live-entry-id"])
        waitForElement(app.staticTexts["locked-entry-id"])

        let defaultEntryIdText = getElementTextById("default-entry-id", app: app)
        let liveEntryIdText = getElementTextById("live-entry-id", app: app)
        let lockedEntryIdText = getElementTextById("locked-entry-id", app: app)

        XCTAssertNotNil(try? ENTRY_ID_TEXT_PATTERN.wholeMatch(in: defaultEntryIdText),
                        "default-entry-id \"\(defaultEntryIdText)\" did not match ENTRY_ID_TEXT_PATTERN")
        XCTAssertNotNil(try? ENTRY_ID_TEXT_PATTERN.wholeMatch(in: liveEntryIdText),
                        "live-entry-id \"\(liveEntryIdText)\" did not match ENTRY_ID_TEXT_PATTERN")
        XCTAssertNotNil(try? ENTRY_ID_TEXT_PATTERN.wholeMatch(in: lockedEntryIdText),
                        "locked-entry-id \"\(lockedEntryIdText)\" did not match ENTRY_ID_TEXT_PATTERN")
    }

    func testDisplaysEntryContentInAllSections() {
        // iOS collapses the RN `*-container` wrapper into the section's
        // accessibility container `*-optimization` — SwiftUI cannot expose a
        // nested accessibility container inside another. The `*-text` assertions
        // below carry the real entry-content verification.
        waitForElement(app.otherElements["default-optimization"])
        waitForElement(app.otherElements["live-optimization"])
        waitForElement(app.otherElements["locked-optimization"])

        let defaultText = getElementTextById("default-text", app: app)
        let liveText = getElementTextById("live-text", app: app)
        let lockedText = getElementTextById("locked-text", app: app)

        // LiveUpdatesEntryDisplay falls back to 'No content' when the resolved
        // entry has no text field — a non-empty text that isn't the fallback
        // proves the SDK fed a real field value.
        XCTAssertGreaterThan(defaultText.count, 0)
        XCTAssertGreaterThan(liveText.count, 0)
        XCTAssertGreaterThan(lockedText.count, 0)
        XCTAssertNotEqual(defaultText, "No content")
        XCTAssertNotEqual(liveText, "No content")
        XCTAssertNotEqual(lockedText, "No content")

        // All three sections wrap the same Contentful entry, so before any
        // identify/toggle/preview-panel actions they MUST resolve to the same
        // variant.
        XCTAssertEqual(defaultText, liveText)
        XCTAssertEqual(defaultText, lockedText)
    }
}

File: implementations/android-sdk/uitests/src/main/kotlin/com/contentful/optimization/uitests/tests/LiveUpdatesTests.kt
package com.contentful.optimization.uitests.tests

import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.platform.app.InstrumentationRegistry
import androidx.test.uiautomator.By
import androidx.test.uiautomator.UiDevice
import com.contentful.optimization.uitests.support.AppLauncher
import com.contentful.optimization.uitests.support.PerTestRule
import com.contentful.optimization.uitests.support.TestHelpers
import com.contentful.optimization.uitests.support.clearProfileState
import org.junit.After
import org.junit.Assert
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.rules.TestRule
import org.junit.runner.RunWith

/**
 * `*-entry-id` Text nodes render "Entry: <sys.id>" where sys.id is alphanumeric.
 * Matching this pattern proves the SDK resolved a real entry rather than an empty/default state.
 */
private val ENTRY_ID_TEXT_PATTERN = Regex("""^Entry: [a-zA-Z0-9]+$""")

@RunWith(AndroidJUnit4::class)
class LiveUpdatesTests {
    @get:Rule
    val rule: TestRule = PerTestRule.create()

    private lateinit var device: UiDevice

    @Before
    fun setUp() {
        device = UiDevice.getInstance(InstrumentationRegistry.getInstrumentation())
        AppLauncher.launchApp(device)
        clearProfileState(device)

        TestHelpers.waitAndTap(device, By.res("live-updates-test-button"))
        // Android app exposes OptimizedEntry via contentDescription ("*-personalization")
        TestHelpers.waitForElement(device, By.desc("default-personalization"), TestHelpers.EXTENDED_TIMEOUT)
    }

    @After
    fun tearDown() {
        val closeButton = device.findObject(By.res("close-live-updates-test-button"))
        if (closeButton != null) TestHelpers.tapElement(device, closeButton)

        device.wait(
            androidx.test.uiautomator.Until.hasObject(By.res("live-updates-test-button")),
            TestHelpers.ELEMENT_TIMEOUT,
        )
    }

    // -------------------------------------------------------------------------
    // Default behavior (locked on first value)
    // -------------------------------------------------------------------------

    @Test
    fun testDefaultDoesNotUpdateOnIdentifyGlobalLiveUpdatesFalse() {
        TestHelpers.waitForElement(device, By.res("default-entry-id"))
        TestHelpers.waitForElement(device, By.res("live-entry-id"))

        val initialDefaultEntryId = TestHelpers.getElementTextById(device, "default-entry-id")
        val initialLiveEntryId = TestHelpers.getElementTextById(device, "live-entry-id")

        TestHelpers.waitAndTap(device, By.res("live-updates-identify-button"))
        TestHelpers.waitForTextEquals(device, "identified-status", "Yes")

        // The live-entry-id section has liveUpdates=true and MUST re-resolve — this is
        // the live-reference that proves the SDK is actually swapping variants.
        TestHelpers.waitForElementText(device, "live-entry-id") { it != initialLiveEntryId }

        // Default section inherits the global setting (off), so the lock must hold.
        TestHelpers.waitForTextEquals(device, "default-entry-id", initialDefaultEntryId)
    }

    // -------------------------------------------------------------------------
    // Global liveUpdates enabled
    // -------------------------------------------------------------------------

    @Test
    fun testGlobalLiveUpdatesEnablesDefaultComponents() {
        TestHelpers.waitAndTap(device, By.res("toggle-global-live-updates-button"))
        TestHelpers.waitForTextEquals(device, "global-live-updates-status", "ON")

        TestHelpers.waitForElement(device, By.res("default-entry-id"))
        val initialDefaultEntryId = TestHelpers.getElementTextById(device, "default-entry-id")

        TestHelpers.waitAndTap(device, By.res("live-updates-identify-button"))
        TestHelpers.waitForTextEquals(device, "identified-status", "Yes")

        // Global=ON means the default section (no per-component prop) MUST re-resolve.
        TestHelpers.waitForElementText(device, "default-entry-id") { it != initialDefaultEntryId }
    }

    @Test
    fun testLockedComponentsIgnoreGlobalLiveUpdates() {
        TestHelpers.waitAndTap(device, By.res("toggle-global-live-updates-button"))
        TestHelpers.waitForTextEquals(device, "global-live-updates-status", "ON")

        TestHelpers.waitForElement(device, By.res("locked-entry-id"))
        TestHelpers.waitForElement(device, By.res("default-entry-id"))

        val initialLockedEntryId = TestHelpers.getElementTextById(device, "locked-entry-id")
        val initialDefaultEntryId = TestHelpers.getElementTextById(device, "default-entry-id")

        TestHelpers.waitAndTap(device, By.res("live-updates-identify-button"))
        TestHelpers.waitForTextEquals(device, "identified-status", "Yes")

        // With global=ON the default section (no per-component prop) MUST re-resolve —
        // the live-reference that proves the SDK is actually swapping variants.
        TestHelpers.waitForElementText(device, "default-entry-id") { it != initialDefaultEntryId }

        // Locked section has liveUpdates=false, so it must stay at its captured id.
        TestHelpers.waitForTextEquals(device, "locked-entry-id", initialLockedEntryId)
    }

    // -------------------------------------------------------------------------
    // Per-component liveUpdates=true
    // -------------------------------------------------------------------------

    @Test
    fun testLiveComponentUpdatesRegardlessOfGlobal() {
        TestHelpers.waitForTextEquals(device, "global-live-updates-status", "OFF")
        TestHelpers.waitForElement(device, By.res("live-entry-id"))

        val initialLiveEntryId = TestHelpers.getElementTextById(device, "live-entry-id")

        TestHelpers.waitAndTap(device, By.res("live-updates-identify-button"))
        TestHelpers.waitForTextEquals(device, "identified-status", "Yes")

        // Per-component liveUpdates=true must override the global=OFF setting.
        TestHelpers.waitForElementText(device, "live-entry-id") { it != initialLiveEntryId }
    }

    // -------------------------------------------------------------------------
    // Per-component liveUpdates=false
    // -------------------------------------------------------------------------

    @Test
    fun testLockedComponentDoesNotUpdateEvenWhenGlobalOn() {
        TestHelpers.waitAndTap(device, By.res("toggle-global-live-updates-button"))
        TestHelpers.waitForTextEquals(device, "global-live-updates-status", "ON")

        TestHelpers.waitForElement(device, By.res("locked-entry-id"))
        TestHelpers.waitForElement(device, By.res("live-entry-id"))

        val initialLockedEntryId = TestHelpers.getElementTextById(device, "locked-entry-id")
        val initialLiveEntryId = TestHelpers.getElementTextById(device, "live-entry-id")

        TestHelpers.waitAndTap(device, By.res("live-updates-identify-button"))
        TestHelpers.waitForTextEquals(device, "identified-status", "Yes")

        // Live section (per-component liveUpdates=true) MUST change — the per-component
        // prop is the path under test: it must override the global=ON setting and keep
        // the locked section stable.
        TestHelpers.waitForElementText(device, "live-entry-id") { it != initialLiveEntryId }
        TestHelpers.waitForTextEquals(device, "locked-entry-id", initialLockedEntryId)
    }

    // -------------------------------------------------------------------------
    // Preview panel simulation
    // -------------------------------------------------------------------------

    @Test
    fun testPreviewPanelEnablesLiveUpdatesForAll() {
        TestHelpers.waitForTextEquals(device, "preview-panel-status", "Closed")
        TestHelpers.waitAndTap(device, By.res("simulate-preview-panel-button"))
        TestHelpers.waitForTextEquals(device, "preview-panel-status", "Open")

        TestHelpers.waitForElement(device, By.res("default-entry-id"))
        TestHelpers.waitForElement(device, By.res("live-entry-id"))
        TestHelpers.waitForElement(device, By.res("locked-entry-id"))

        val initialDefaultEntryId = TestHelpers.getElementTextById(device, "default-entry-id")
        val initialLiveEntryId = TestHelpers.getElementTextById(device, "live-entry-id")
        val initialLockedEntryId = TestHelpers.getElementTextById(device, "locked-entry-id")

        TestHelpers.waitAndTap(device, By.res("live-updates-identify-button"))
        TestHelpers.waitForTextEquals(device, "identified-status", "Yes")

        // While the preview panel is open, the SDK forces shouldLiveUpdate=true for ALL
        // sections, including the per-component liveUpdates=false one.
        TestHelpers.waitForElementText(device, "default-entry-id") { it != initialDefaultEntryId }
        TestHelpers.waitForElementText(device, "live-entry-id") { it != initialLiveEntryId }
        TestHelpers.waitForElementText(device, "locked-entry-id") { it != initialLockedEntryId }
    }

    // -------------------------------------------------------------------------
    // Screen controls
    // -------------------------------------------------------------------------

    @Test
    fun testToggleGlobalLiveUpdates() {
        TestHelpers.waitForTextEquals(device, "global-live-updates-status", "OFF")
        TestHelpers.waitAndTap(device, By.res("toggle-global-live-updates-button"))
        TestHelpers.waitForTextEquals(device, "global-live-updates-status", "ON")
        TestHelpers.waitAndTap(device, By.res("toggle-global-live-updates-button"))
        TestHelpers.waitForTextEquals(device, "global-live-updates-status", "OFF")
    }

    @Test
    fun testTogglePreviewPanel() {
        TestHelpers.waitForTextEquals(device, "preview-panel-status", "Closed")
        TestHelpers.waitAndTap(device, By.res("simulate-preview-panel-button"))
        TestHelpers.waitForTextEquals(device, "preview-panel-status", "Open")
        TestHelpers.waitAndTap(device, By.res("simulate-preview-panel-button"))
        TestHelpers.waitForTextEquals(device, "preview-panel-status", "Closed")
    }

    @Test
    fun testIdentifyAndReset() {
        TestHelpers.waitForTextEquals(device, "identified-status", "No")
        TestHelpers.waitAndTap(device, By.res("live-updates-identify-button"))
        TestHelpers.waitForTextEquals(device, "identified-status", "Yes")
        TestHelpers.waitAndTap(device, By.res("live-updates-reset-button"))
        TestHelpers.waitForTextEquals(device, "identified-status", "No")
    }

    // -------------------------------------------------------------------------
    // Three Optimization sections display
    // -------------------------------------------------------------------------

    @Test
    fun testDisplaysAllThreeOptimizationEntrySections() {
        // Android app exposes OptimizedEntry via contentDescription ("*-personalization").
        // iOS uses "*-optimization". Both refer to the same SDK-rendered containers.
        TestHelpers.waitForElement(device, By.desc("default-personalization"), TestHelpers.ELEMENT_TIMEOUT)
        TestHelpers.waitForElement(device, By.desc("live-personalization"), TestHelpers.ELEMENT_TIMEOUT)
        TestHelpers.waitForElement(device, By.desc("locked-personalization"), TestHelpers.ELEMENT_TIMEOUT)

        TestHelpers.waitForElement(device, By.res("default-entry-id"))
        TestHelpers.waitForElement(device, By.res("live-entry-id"))
        TestHelpers.waitForElement(device, By.res("locked-entry-id"))

        val defaultEntryIdText = TestHelpers.getElementTextById(device, "default-entry-id")
        val liveEntryIdText = TestHelpers.getElementTextById(device, "live-entry-id")
        val lockedEntryIdText = TestHelpers.getElementTextById(device, "locked-entry-id")

        Assert.assertTrue(
            "default-entry-id \"$defaultEntryIdText\" did not match ENTRY_ID_TEXT_PATTERN",
            ENTRY_ID_TEXT_PATTERN.matches(defaultEntryIdText),
        )
        Assert.assertTrue(
            "live-entry-id \"$liveEntryIdText\" did not match ENTRY_ID_TEXT_PATTERN",
            ENTRY_ID_TEXT_PATTERN.matches(liveEntryIdText),
        )
        Assert.assertTrue(
            "locked-entry-id \"$lockedEntryIdText\" did not match ENTRY_ID_TEXT_PATTERN",
            ENTRY_ID_TEXT_PATTERN.matches(lockedEntryIdText),
        )
    }

    @Test
    fun testDisplaysEntryContentInAllSections() {
        TestHelpers.waitForElement(device, By.res("default-container"))
        TestHelpers.waitForElement(device, By.res("live-container"))
        TestHelpers.waitForElement(device, By.res("locked-container"))

        val defaultText = TestHelpers.getElementTextById(device, "default-text")
        val liveText = TestHelpers.getElementTextById(device, "live-text")
        val lockedText = TestHelpers.getElementTextById(device, "locked-text")

        Assert.assertTrue("default-text should be non-empty", defaultText.isNotEmpty())
        Assert.assertTrue("live-text should be non-empty", liveText.isNotEmpty())
        Assert.assertTrue("locked-text should be non-empty", lockedText.isNotEmpty())
        Assert.assertNotEquals("default-text should not be 'No content'", "No content", defaultText)
        Assert.assertNotEquals("live-text should not be 'No content'", "No content", liveText)
        Assert.assertNotEquals("locked-text should not be 'No content'", "No content", lockedText)
        // Before any identify/toggle/preview-panel action all three sections wrap the
        // same Contentful entry and MUST resolve to the same variant text.
        Assert.assertEquals("default-text and live-text should match", defaultText, liveText)
        Assert.assertEquals("default-text and locked-text should match", defaultText, lockedText)
    }
}
